### ISSUE TITLE
Delta does not allow UpdatableProperties to be changed

### ISSUE THREAD

**Issue by dxrdxr** [OData/WebApi#2478]
The framework creates a Delta for Patch() but in many scenarios the controller wants to protect from the client
over-posting. When searching StackOverflow questions such as [this](https://stackoverflow.com/questions/28248785/odata-delta-patch-security) show that the solutions are very complex.

### Assemblies affected

OData WebApi lib 7.x

### Reproduce steps

Use the sample controller's Patch code and try to restrict the properties that are updated.

### Expected result

Only update the properties that are specified by the controller.

### Actual result

All modified properties are updated.

### Additional detail

When a Delta is constructed, it is possible to specify the `updatableProperties` parameter to restrict which properties can be set in the Delta, and also limit which properties will be copied on `Patch` and `Put`. However, once the Delta is constructed the list cannot be changed. Furthermore, because the framework constructs the Delta it cannot be specified by the controller.

Workarounds on the Internet typically involve one of the following:

- Reconstructing the Delta through extensions to find all changed properties and only copying the selected
- Replacing the Put/Patch method with an equivalent method that explicitly does the copying

A much simpler solution is to expose the property:

```
public IEnumerable<string> UpdatableProperties { get; set; } 
```

The getter simply returns the existing `_updatableProperties`.

The setter replaces `_updatableProperties` with the `value` and additionally updates the `_changedProperties` to remove any properties that are no longer in the `_updatableProperties`.

Subsequently the Put() and Patch() methods work as expected, only the specified properties are updated.

Example:
```
public void Patch(int productId, Delta<Product> product)
{
    var productFromDb = // get product from db by productId
    // Restrict the updatable properties to prevent over posting
    product.UpdatableProperties =  new[] { "SafeProperty1", "SafeProperty2"};
    product.Patch(productFromDb);
    // some other stuff
}
```

**Comment by habbes** [comment]
Hi @dxrdxr what if you create an overload of `Patch` and `Put` that accept a list of "safe" properties?

Would you also consider contributing a PR for this feature?

**Comment by dxrdxr** [comment]
I don't think an overload is a good solution @habbes for a few reasons:

1. It should also include an overload for `CopyChangedValues` which is the functional equivalent of `Patch`
2. Likewise it should also include an overload for `CopyUnchangedValues` which is half of `Put`
3. Following this forward, the caller of `GetChangedPropertyNames` and `GetUnchangedPropertyNames` will have to filter results by the list of "safe" properties. Currently most of the hacks around this problem is full of filtering which minimizes the effectiveness of this approach.
4. Trying to cleanly implement some or all of 1-3 is rather messy internally.
5. The biggest problem is that using an overload will not work for Delta Nested Resources. The `Patch` mechanics for nested resources is to recursively call `CopyChangedValues` which will need to be passed a list of "safe" properties for the nested resource. This will require us to create a new syntax to specify the list of properties for nesting. The straightforward approach is to use a path such as "nested1/nested2/nested3/safeproperty". This again is going to add a lot of complexity to the implementation.

My proposed solution (which I have implemented already with minimal changes, _PR is available_), solves 1-4 cleanly.

For #5, my proposal just works. The semantics for nested resources is that the caller who is making changes to the resources is setting property with an `IDelta` (or explicitly calling `TrySetPropertyValue`) containing the changes. So using my approach to restrict properties follows a symmetric approach:

```
public void Patch(int productId, Delta<Product> product)
{
    var productFromDb = // get product from db by productId
    // Restrict the updatable properties to prevent over posting
    dynamic idelta = product.nested1;
    idelta.UpdatableProperties = new[] { "Nested1SafeProperty"};
    product.UpdatableProperties =  new[] { "SafeProperty1", "SafeProperty2"};
    product.Patch(productFromDb);
    // some other stuff
}
```

What I have proposed is a minimal change to the existing interface, the internal implementation is simple and small, and the use case for nested resources follows the existing model (explicitly set at each level) and requires no extra implementation changes.

PR available.
